### ISSUE TITLE
Memory Management

### ISSUE THREAD

**Issue by FiV0** [xtdb/crux#3891]
The buffer pool (BP henceforth) deals with reading and writing files. The ticket concerns the reading side of the BP. As XT is an append only system the reading of pages does not interfere with the writing of pages. The buffer pool also deals with management of disk caches which we will not be concerned with here. In the following record batch and page is somewhat used interchangeably. 

The memory management (read-side) as it currently stands:
- queries and also DML (if it is not a simple put statement) require to read from certain files when scanning
- the BP memory maps the file and scan then reads the footer and loads a page accordingly
- If multiple queries take out the same buffer the ref count is increased and only after the last query closes the buffer pool may decide to evict the mmap buffer from the cache.
- The upside of this approach is that due to memory-mapped files we do not have to deal with memory management too closely.
- The downside is that due to memory mapping we actually have very little good information of how much memory is used and are somewhat at the whim of the OS paging policies (which might not be very good for databases).
- The BP is also currently non blocking (as blocking introduces all kinds of other issues, more on that below). This means that files are memory mapped even if the maximum cache size is exceeded (see for example https://github.com/xtdb/xtdb/issues/3872). 

The following is an outline of some changes to the buffer pool to potential deal with the memory management finer grained ourselves:
- Allow for maximum size record batch (not necessarily in intermediate operators (join, sort etc..)) but for scan, ie. on disk.
  - The preliminary work would be to get some estimate on the size in bytes of a Arrow vector/relation.
- A query then has an upper bound on how much memory it will be using from the buffer pool (roughly the number of scans)
- Based on this upper bound a query gets queued until the buffer pool can serve the required number of record batches.
   - Issue: A query in the indexer should never be blocked.
- This introduces a finer hierarchy in the buffer pool. The unit of memory handed out is then a record batch + header rather than a whole memory mapped file. The record batch memory is than reused for subsequent record batch loading. There are two options here:
   - The buffer pool only deals in record batches and can hand out the same record batch to multiple queries. This means the buffer pool is mainly concerned with reusing the memory for handed back record batches.
   - The buffer pool hands out something that can hold a record batch and the footer and the queries (scan) themselves deal with loading the appropriate data into the record batch. This kind of closes the door on sharing pages as any query loads whatever it needs into the slot for the record batch.
- The blocking of BP requests might be a Pandora's box. Do we need to be able to preempt other BP usage for high priority requests?  Should there be timeouts for BP requests? Do we then need a query schedular?
   
Potential issues:
- The meta file (trie file) is currently always one record batch, so we might also need to buffer that if the tries become large. As it currently stands this likely secondary, as these files are << compared to data files. 
- An indexing thread query should essentially never be blocked. Do we need to preempt other queries? Do we need priorities for requests to the buffer pool?

Stuff to maybe look into:
- content-addressable file-systems/caches
- `O_Direct`
- `fadvise`

Other databases approaches:
- timeout and freeup the resources - this kind of assumes this scenario does not happen very often
- deadlock detection
- query schedular
- lock the the whole db (sqllite) - I don't think this is an option

Out of scope:
- batching and memory mangement in operators

Progress
- [x] Passed to memory-mapped record batches in #3919

**Comment by FiV0** [comment]
We passed to memory-mapped record batches in #3919.

**Comment by FiV0** [comment]
### State of direct vs native
- Goal: Count all non heap memory of XT under one counter. 
- Currently we are using the arrow `memory-netty` dependency. There is also `memory-unsafe` which (I assume) calls `UNSAFE` directly if available. This is to my understanding doing pretty much the same as netty but with less wrapping.
- We are (currently) in certain places calling `ByteBuffer/allocateDirect` (for example when creating views over written data). This gets accounted for in the `direct` memory pool of `java.nio.Bits`. If this counter goes above the limit set by `XX:MaxDirectMemorySize=1g` `ByteBuffer/allocateDirect` fails.
- When we are writing to arrow (compactor, ra operaters, etc...) vectors might need to grow. This allocation happens through netty itself (and currently bypasses) the `MaxDirectMemorySize` limit. The allocated memory can be observed with `(io.netty.util.internal.PlatformDependent/usedDirectMemory)`
- Netty itself has a couple of options that influence through what mechanism it allocates memory.
	- `-Dio.netty.tryReflectionSetAccessible=true` - after jdk9 this is needed so that netty can construct DirectByteBuffers without a cleaner. This is currently enabled in our case.
	- `-Dio.netty.maxDirectMemory` - copying from the comments from the netty source + additional comments.
		- \* <  0  - Don't use cleaner, and inherit max direct memory from java. In this case the "practical max direct memory" would be 2 * max memory as defined by the JDK. Netty and real JVM direct memory get counted seperately.
		- \* == 0  - Use cleaner, Netty will not enforce max memory, and instead will defer to the JDK.  This essentially uses `Buffer/allocateDirect`. Memory is freed when GCing.
		- \* >  0  - Don't use cleaner. This will limit Netty's total direct memory (note: that JDK's direct memory limit is independent of this). This is the same as the first option but you can specify netty's and the direct memory separately.

Options: 
1. Use `-Dio.netty.maxDirectMemory=0` . It's no longer possible to free buffers ourselfs. The GC does it for us. Netty and direct buffers are the same thing as netty does `ByteBuffer/allocateDirect`.
2. Move everything to netty. Use `-Dio.netty.maxDirectMemory=2000000000`  I think the main issue is getting hold of the underlying netty pools. But it is possible to create two independent pools and the limit is constraint across all of them. The JVM would still be able to call `allocateDirect` elsewhere but we would avoid it. Test for this below.
3. Keep default JVM flags to `MaxDirectMemorySize`. `io.netty.maxDirectMemory` defaults to -1 and hence the JVM `MaxDirectMemorySize` is used. We need to make sure that we are no longer using `ByteBuffer/allocateDirect` anywhere else. We use a netty pool in the memory cache and for the memory buffer pool.

Test case for the second option (with `-Dio.netty.maxDirectMemory=500000000`).
```clj
  (def alloc (RootAllocator.))

  (def request-size (* 256 1024 1024))

  ;; allocation when growing
  (def arrow-buf (.buffer alloc request-size))

  (import '(io.netty.buffer PooledByteBufAllocator))

  (def netty-alloc (PooledByteBufAllocator. true))

  ;; allocation when dealing with it ourselves
  ;; fails
  (def netty-buf (.directBuffer netty-alloc request-size request-size))

  (io.netty.util.internal.PlatformDependent/maxDirectMemory)
  ;; => 500000000

  (io.netty.util.internal.PlatformDependent/usedDirectMemory)
  ;; => 268435456
```

**Comment by FiV0** [comment]
Going with option 3. for now.
